File: prototype/rails/src/test/domain/orders/order_status_test.rb
require "test_helper"

module Domain
  module Orders
    class OrderStatusTest < ActiveSupport::TestCase
      test "every status has a transition list" do
        assert_equal OrderStatus::ALL.sort, OrderStatus::TRANSITIONS.keys.sort
      end

      test "verifying an email opens payment" do
        assert OrderStatus.can_transition?(OrderStatus::PENDING_VERIFICATION, OrderStatus::AWAITING_PAYMENT)
      end

      test "an order awaiting payment is paid or fails" do
        assert OrderStatus.can_transition?(OrderStatus::AWAITING_PAYMENT, OrderStatus::PAID)
        assert OrderStatus.can_transition?(OrderStatus::AWAITING_PAYMENT, OrderStatus::PAYMENT_FAILED)
      end

      test "a guest cannot pay before verifying" do
        refute OrderStatus.can_transition?(OrderStatus::PENDING_VERIFICATION, OrderStatus::PAID)
      end

      test "a failed payment retries" do
        assert OrderStatus.can_transition?(OrderStatus::PAYMENT_FAILED, OrderStatus::PAID)
      end

      test "a retry that is declined again stays where it was" do
        assert OrderStatus.can_transition?(OrderStatus::PAYMENT_FAILED, OrderStatus::PAYMENT_FAILED)
      end

      test "a paid order ships whole or in part" do
        assert OrderStatus.can_transition?(OrderStatus::PAID, OrderStatus::SHIPPED)
        assert OrderStatus.can_transition?(OrderStatus::PAID, OrderStatus::PARTIALLY_SHIPPED)
      end

      test "a delivered order goes nowhere" do
        assert_empty OrderStatus::TRANSITIONS.fetch(OrderStatus::DELIVERED)
      end

      test "a cancelled order goes nowhere" do
        assert_empty OrderStatus::TRANSITIONS.fetch(OrderStatus::CANCELLED)
      end

      test "a paid order cannot be paid twice" do
        error = assert_raises(TransitionError) { OrderStatus.transition(OrderStatus::PAID, OrderStatus::PAID) }
        assert_equal "An order cannot move from paid to paid.", error.message
      end

      test "a verified purchaser places an order that awaits payment" do
        assert_equal OrderStatus::AWAITING_PAYMENT, OrderStatus.for_placement(verified_purchaser)
      end

      test "a guest places an order that awaits verification" do
        assert_equal OrderStatus::PENDING_VERIFICATION, OrderStatus.for_placement(guest_purchaser)
      end

      test "verification moves an order that was waiting on it" do
        assert_equal OrderStatus::AWAITING_PAYMENT, OrderStatus.after_verification(OrderStatus::PENDING_VERIFICATION)
      end

      test "verification leaves every other status alone" do
        assert_equal OrderStatus::PAID, OrderStatus.after_verification(OrderStatus::PAID)
        assert_equal OrderStatus::AWAITING_PAYMENT, OrderStatus.after_verification(OrderStatus::AWAITING_PAYMENT)
      end

      test "an approved card pays the order" do
        assert_equal OrderStatus::PAID, OrderStatus.from_card_decision(Payments::CardDecision.approved("4242"))
      end

      test "a declined card fails the payment" do
        decision = Payments::CardDecision.declined("0002", "generic_decline")

        assert_equal OrderStatus::PAYMENT_FAILED, OrderStatus.from_card_decision(decision)
      end

      test "an order whose fulfillments all delivered is delivered" do
        assert_equal OrderStatus::DELIVERED, OrderStatus.from_fulfillments(%w[delivered delivered])
      end

      test "an order whose fulfillments all departed is shipped" do
        assert_equal OrderStatus::SHIPPED, OrderStatus.from_fulfillments(%w[shipped delivered])
      end

      test "an order with one fulfillment still in the studio is partially shipped" do
        assert_equal OrderStatus::PARTIALLY_SHIPPED, OrderStatus.from_fulfillments(%w[delivered awaiting_shipment])
      end

      test "an order whose fulfillments all await shipment is paid" do
        assert_equal OrderStatus::PAID, OrderStatus.from_fulfillments(%w[awaiting_shipment awaiting_shipment])
      end

      test "an order rolls up from at least one fulfillment" do
        assert_raises(ArgumentError) { OrderStatus.from_fulfillments([]) }
      end

      private

      def verified_purchaser
        Purchaser.new(id: 1, email: "ada@example.test", email_verified_at: Time.utc(2026, 8, 19))
      end

      def guest_purchaser
        Purchaser.new(id: 2, email: "guest@example.test", email_verified_at: nil)
      end
    end
  end
end

File: prototype/node/src/app/core/orders/order-status.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { BrokenContractError } from '../defect.ts'
import { refused } from '../refusal.ts'
import {
  ORDER_STATUSES,
  ORDER_STATUS_TRANSITIONS,
  canTransitionOrder,
  transitionOrder,
  orderMovedTo,
  orderTransitionRefusalCopy,
  orderStatusForPlacement,
  orderStatusAfterVerification,
  orderStatusFromCardDecision,
  orderStatusFromFulfillments,
  isCancellable,
} from './order-status.ts'
import { approvedCard, declinedCard } from '../payments/card-decision.ts'

test('every status has a transition list', () => {
  assert.deepEqual(Object.keys(ORDER_STATUS_TRANSITIONS).sort(), [...ORDER_STATUSES].sort())
})

test('verifying an email opens payment', () => {
  assert.equal(canTransitionOrder('pending_verification', 'awaiting_payment'), true)
})

test('an order awaiting payment is paid or fails', () => {
  assert.equal(canTransitionOrder('awaiting_payment', 'paid'), true)
  assert.equal(canTransitionOrder('awaiting_payment', 'payment_failed'), true)
})

test('a guest cannot pay before verifying', () => {
  assert.equal(canTransitionOrder('pending_verification', 'paid'), false)
})

test('a failed payment retries', () => {
  assert.equal(canTransitionOrder('payment_failed', 'paid'), true)
})

test('a retry that is declined again stays where it was', () => {
  assert.equal(canTransitionOrder('payment_failed', 'payment_failed'), true)
})

test('a paid order ships whole or in part', () => {
  assert.equal(canTransitionOrder('paid', 'shipped'), true)
  assert.equal(canTransitionOrder('paid', 'partially_shipped'), true)
})

test('a delivered order goes nowhere but a refund', () => {
  assert.deepEqual(ORDER_STATUS_TRANSITIONS.delivered, ['refunded'])
})

test('a refunded order goes nowhere', () => {
  assert.deepEqual(ORDER_STATUS_TRANSITIONS.refunded, [])
})

test('every paid stage reaches refunded', () => {
  assert.equal(canTransitionOrder('paid', 'refunded'), true)
  assert.equal(canTransitionOrder('partially_shipped', 'refunded'), true)
  assert.equal(canTransitionOrder('shipped', 'refunded'), true)
  assert.equal(canTransitionOrder('delivered', 'refunded'), true)
})

test('an unpaid order is cancelled rather than refunded', () => {
  assert.equal(canTransitionOrder('pending_verification', 'refunded'), false)
  assert.equal(canTransitionOrder('awaiting_payment', 'refunded'), false)
  assert.equal(canTransitionOrder('payment_failed', 'refunded'), false)
})

test('a cancelled order cannot be paid', () => {
  assert.equal(canTransitionOrder('cancelled', 'paid'), false)
})

test('a cancelled order goes nowhere', () => {
  assert.deepEqual(ORDER_STATUS_TRANSITIONS.cancelled, [])
})

test('transition returns the next status', () => {
  assert.deepEqual(transitionOrder('awaiting_payment', 'paid'), { outcome: 'allowed', status: 'paid' })
})

test('a paid order cannot be paid twice', () => {
  assert.deepEqual(
    transitionOrder('paid', 'paid'),
    refused('illegal_transition', { status_from: 'paid', status_to: 'paid' }),
  )
})

test('orderMovedTo returns the status for a legal move', () => {
  assert.equal(orderMovedTo('awaiting_payment', 'paid'), 'paid')
})

test('orderMovedTo throws for a move the table does not allow', () => {
  assert.throws(
    () => orderMovedTo('paid', 'paid'),
    (error: unknown) =>
      error instanceof BrokenContractError &&
      error.reason === 'illegal_transition' &&
      error.message === 'An order cannot move from paid to paid.' &&
      JSON.stringify(error.data) === JSON.stringify({ status_from: 'paid', status_to: 'paid' }),
  )
})

test('orderTransitionRefusalCopy words the illegal move from the refusal data', () => {
  const refusal = refused('illegal_transition', { status_from: 'paid', status_to: 'paid' } as const)

  assert.equal(orderTransitionRefusalCopy(refusal), 'An order cannot move from paid to paid.')
})

test('orderTransitionRefusalCopy renders the refusal data, not a status a route read before the race', () => {
  // The action's refusal carries the status as of the write; the route's earlier
  // row read is stale by the time a concurrent move lands first.
  const routeReadBeforeTheRace = 'awaiting_payment'
  const refusal = refused('illegal_transition', { status_from: 'paid', status_to: 'shipped' } as const)

  const sentence = orderTransitionRefusalCopy(refusal)

  assert.match(sentence, /paid/)
  assert.doesNotMatch(sentence, new RegExp(routeReadBeforeTheRace))
})

test('a verified purchaser places an order that awaits payment', () => {
  assert.equal(orderStatusForPlacement(true), 'awaiting_payment')
})

test('a guest places an order that awaits verification', () => {
  assert.equal(orderStatusForPlacement(false), 'pending_verification')
})

test('verification moves an order that was waiting on it', () => {
  assert.equal(orderStatusAfterVerification('pending_verification'), 'awaiting_payment')
})

test('verification leaves every other status alone', () => {
  assert.equal(orderStatusAfterVerification('paid'), 'paid')
  assert.equal(orderStatusAfterVerification('awaiting_payment'), 'awaiting_payment')
})

test('an approved card pays the order', () => {
  assert.equal(orderStatusFromCardDecision(approvedCard('4242')), 'paid')
})

test('a declined card fails the payment', () => {
  assert.equal(orderStatusFromCardDecision(declinedCard('0002', 'generic_decline')), 'payment_failed')
})

test('an order whose fulfillments all delivered is delivered', () => {
  assert.equal(orderStatusFromFulfillments(['delivered', 'delivered']), 'delivered')
})

test('an order whose fulfillments all departed is shipped', () => {
  assert.equal(orderStatusFromFulfillments(['shipped', 'delivered']), 'shipped')
})

test('an order with one fulfillment still in the studio is partially shipped', () => {
  assert.equal(orderStatusFromFulfillments(['delivered', 'awaiting_shipment']), 'partially_shipped')
})

test('an order whose fulfillments all await shipment is paid', () => {
  assert.equal(orderStatusFromFulfillments(['awaiting_shipment', 'awaiting_shipment']), 'paid')
})

test('an order whose fulfillments are all reversed is refunded', () => {
  assert.equal(orderStatusFromFulfillments(['declined']), 'refunded')
  assert.equal(orderStatusFromFulfillments(['refunded']), 'refunded')
  assert.equal(orderStatusFromFulfillments(['declined', 'refunded']), 'refunded')
})

test('a mixed order rolls up from its live fulfillments only', () => {
  assert.equal(orderStatusFromFulfillments(['shipped', 'declined']), 'shipped')
  assert.equal(orderStatusFromFulfillments(['delivered', 'refunded']), 'delivered')
  assert.equal(orderStatusFromFulfillments(['awaiting_shipment', 'declined']), 'paid')
  assert.equal(orderStatusFromFulfillments(['shipped', 'awaiting_shipment', 'declined']), 'partially_shipped')
})

test('an order rolls up from at least one fulfillment', () => {
  assert.throws(() => orderStatusFromFulfillments([]), RangeError)
})

test('isCancellable names the three statuses a customer can cancel from', () => {
  assert.equal(isCancellable('pending_verification'), true)
  assert.equal(isCancellable('awaiting_payment'), true)
  assert.equal(isCancellable('payment_failed'), true)
})

test('isCancellable refuses a paid order', () => {
  assert.equal(isCancellable('paid'), false)
})
